File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter17/basic/BasicDesignPatterns.kt
package com.devlokos.masteringkotlin.chapters.chapter17.basic

/**
 * Capítulo 17 — Patrones de Diseño.
 * Nivel BÁSICO: Singleton — una única instancia compartida (Kotlin: declaración object).
 */
fun main() {
    runBasicDesignPatternExamples()
}

/** Singleton: 'object' en Kotlin crea una única instancia. */
object AppConfig {
    val appName = "Mastering Kotlin"
    val version = "1.0"
}

fun runBasicDesignPatternExamples() {
    println("AppConfig.appName = ${AppConfig.appName}")
    println("AppConfig.version = ${AppConfig.version}")
    println("Same instance: ${AppConfig === AppConfig}")
}

File: MasteringDart/lib/chapters/chapter17/basic/basic_design_patterns.dart
// ignore_for_file: avoid_print
//
//  basic_design_patterns.dart
//  MasteringDart
//
//  Capítulo 17 — Patrones de diseño.
//  Nivel BÁSICO: Singleton — una única instancia compartida (factory + static _instance).
//
//  Ejecutar con: dart run lib/chapters/chapter17/basic/basic_design_patterns.dart

class AppConfig {
  AppConfig._();
  static final AppConfig _instance = AppConfig._();
  factory AppConfig() => _instance;

  final String appName = 'Mastering Dart';
  final String version = '1.0';
}

void runBasicDesignPatternExamples() {
  final a = AppConfig();
  final b = AppConfig();
  print('AppConfig().appName = ${a.appName}');
  print('AppConfig().version = ${a.version}');
  print('Same instance: ${identical(a, b)}');
}

void main() {
  runBasicDesignPatternExamples();
}
